### PR TITLE
Allow users to disable TLS for HTTP in the Elasticsearch spec

### LINKED ISSUE #1615: Implement `http.tls.selfSignedCertificate.disabled` for Elasticsearch
The corresponding option is currently supported for APM and Kibana only. We should implement it for Elasticsearch as well: 

* to be consistent across all objects
* to support use cases where users want to have unencrypted client access via HTTP

Implications:
* readiness probes need to be protocol aware
* our internal Elasticsearch client needs to be protocol aware
* client construction needs to be able to deal with both cases

### PR DESCRIPTION
**Pull request by pebrc**
Fixes #1615 

* adds the ability to turn off TLS on the HTTP layer
* this creates some complications for upgrades from HTTP->HTTPS and vice versa
   * refactors ES URL calculation to use single pod requests while migrating from/to TLS
   * this in turn makes some of our e2e logic more complicated because we now have to deal with individual pods during the transition and tests need to expect that pods can go away
   * I opted for making the e2e tests a bit more flexible and am recreating the ES client for each step during the mutation test steps to catch these cases which seemed preferable over make this edge case very prominent and introducing  a separate service just the handle the outlier case of users wanting to switch to/from TLS

### CODE REVIEW

File: pkg/apis/common/v1alpha1/common_test.go [hunk 1]
```diff
@@ -67,3 +67,54 @@ func TestTLSOptions_Enabled(t *testing.T) {
...
+				},
+			},
+			want: "https",
+		},
+		{
+			name: "disabled",
+			fields: fields{
+				TLS: TLSOptions{
+					SelfSignedCertificate: &SelfSignedCertificate{
+						Disabled: true,
```
Comment (barkbay) [inline pkg/apis/common/v1alpha1/common_test.go]: nit: what about a test with `SelfSignedCertificate` disabled but with a custom cert ?
```golang
		{
			name: "enabled: custom certs and self-signed disabled",
			fields: fields{
				TLS: TLSOptions{
					SelfSignedCertificate: &SelfSignedCertificate{
						Disabled: true,
					},
					Certificate: SecretRef{
						SecretName: "my-custom-certs",
					},
				},
			},
			want: "https",
		},
```

File: pkg/controller/elasticsearch/driver/driver.go [hunk 2]
```diff
@@ -229,7 +229,7 @@ func (d *defaultDriver) newElasticsearchClient(
 	v version.Version,
 	caCerts []*x509.Certificate,
 ) esclient.Client {
-	url := fmt.Sprintf("https://%s.%s.svc:%d", service.Name, service.Namespace, network.HTTPPort)
+	url := fmt.Sprintf("%s://%s.%s.svc:%d", d.ES.Spec.HTTP.Scheme(), service.Name, service.Namespace, network.HTTPPort)
```
Comment (sebgl) [inline pkg/controller/elasticsearch/driver/driver.go]: I think we need to be smarter if we want to handle TLS <-> non-TLS migrations. Rolling upgrade would probably fail since we cannot request the cluster.

Comment (pebrc) [reply pkg/controller/elasticsearch/driver/driver.go]: Indeed!

This PR is still relatively small, we could try to address it right here, wdyt?

Comment (sebgl) [reply pkg/controller/elasticsearch/driver/driver.go]: up to you :)

File: test/e2e/test/elasticsearch/checks_data.go [hunk 3]
```diff
@@ -67,7 +68,7 @@ func (dc *DataIntegrityCheck) Init() error {
 	if err != nil {
 		return err
 	}
-	resp, err := dc.client.Request(context.Background(), indexCreation)
+	resp, err := esClient.Request(context.Background(), indexCreation)
 	defer resp.Body.Close() // nolint
```
Comment (barkbay) [inline test/e2e/test/elasticsearch/checks_data.go]: I need to investigate more but I had this error while running the e2e tests _(maybe it comes from my setup)_:

```
  === RUN   TestMutationHTTPToHTTPS/Add_some_data_to_the_cluster_before_starting_the_mutation
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x1069576]

goroutine 340 [running]:
testing.tRunner.func1(0xc00037a500)
	/usr/local/go/src/testing/testing.go:792 +0x387
panic(0x119f7a0, 0x1f21720)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/elastic/cloud-on-k8s/test/e2e/test/elasticsearch.(*DataIntegrityCheck).Init(0xc0007261c0, 0x0, 0x0)
	/go/src/github.com/elastic/cloud-on-k8s/test/e2e/test/elasticsearch/checks_data.go:72 +0x2f6
github.com/elastic/cloud-on-k8s/test/e2e/test/elasticsearch.Builder.MutationTestSteps.func1(0xc00037a500)
	/go/src/github.com/elastic/cloud-on-k8s/test/e2e/test/elasticsearch/steps_mutation.go:47 +0x98
testing.tRunner(0xc00037a500, 0xc0000eb720)
	/usr/local/go/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
```

I think `defer resp.Body.Close()` should be done after the error check anyway.

**Comment by pebrc** [comment]
Good find, I forgot about this. When there is a protocol mismatch you can
get a nil response

File: pkg/controller/elasticsearch/services/services.go [hunk 4]
```diff
@@ -90,3 +92,23 @@ func GetExternalService(c k8s.Client, es v1alpha1.Elasticsearch) (corev1.Service
...
+		scheme, exists := p.Labels[label.HTTPSchemeLabelName]
+		if exists && scheme != es.Spec.HTTP.Scheme() {
+			// scheme in existing pods does not match scheme in spec, user toggled HTTP(S)
+			schemeChange = true
+		}
+	}
+	if schemeChange {
+		// switch to sending requests directly to a random pod instead of going through the service
+		randomPod := pods[rand.Intn(len(pods))]
+		return fmt.Sprintf("%s://%s.%s:%d", randomPod.Labels[label.HTTPSchemeLabelName], randomPod.Name, randomPod.Namespace, network.HTTPPort)
```
Comment (barkbay) [inline pkg/controller/elasticsearch/services/services.go]: Got this error:
```
    --- FAIL: TestMutationHTTPToHTTPS/Add_some_data_to_the_cluster_before_starting_the_mutation (0.04s)
        require.go:794:
            	Error Trace:	steps_mutation.go:47
            	Error:      	Received unexpected error:
            	            	Put http://test-mutation-http-to-https-rzdh-es-masterdata-0.e2e-68x09-mercury:9200/data-integrity-check: dial tcp: lookup test-mutation-http-to-https-rzdh-es-masterdata-0.e2e-68x09-mercury on 10.96.0.10:53: no such host
            	Test:       	TestMutationHTTPToHTTPS/Add_some_data_to_the_cluster_before_starting_the_mutation
FAIL
FAIL	github.com/elastic/cloud-on-k8s/test/e2e/es	65.316s
```

IIUC the [statefulset documentation](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id) we need a headless service between the Pod's name and the namespace to be able to resolve a POD name:

```
[elastic@e2e-68x09-mercury-ns-operator-0 ~]$ ping test-mutation-http-to-https-rzdh-es-masterdata-0.test-mutation-http-to-https-rzdh-es-masterdata.e2e-68x09-mercury
PING test-mutation-http-to-https-rzdh-es-masterdata-0.test-mutation-http-to-https-rzdh-es-masterdata.e2e-68x09-mercury.svc.cluster.local (10.244.2.4) 56(84) bytes of data.
64 bytes from 10-244-2-4.test-mutation-http-to-https-rzdh-es-http.e2e-68x09-mercury.svc.cluster.local (10.244.2.4): icmp_seq=1 ttl=62 time=0.077 ms
```

Comment (pebrc) [reply pkg/controller/elasticsearch/services/services.go]: I think you are right 🥇  Interestingly this error does not occur when running the e2e tests locally, with our port-forwarding tooling which is what I did 😞

File: test/e2e/test/elasticsearch/steps_mutation.go [hunk 5]
```diff
@@ -77,9 +75,9 @@ func (b Builder) MutationTestSteps(k *test.K8sClient) test.StepList {
 			},
 			test.Step{
 				Name: "Data added initially should still be present",
-				Test: func(t *testing.T) {
-					require.NoError(t, dataIntegrityCheck.Verify())
-				},
+				Test: test.Eventually(func() error { // nolint
+					return dataIntegrityCheck.Verify()
```
Comment (pebrc) [inline test/e2e/test/elasticsearch/steps_mutation.go]: This should be `test.Eventually(dataIntegrityCheck.Verify)` but that does not work and panics which I don't fully understand yet

File: pkg/controller/elasticsearch/services/services.go [hunk 6]
```diff
@@ -90,3 +92,27 @@ func GetExternalService(c k8s.Client, es v1alpha1.Elasticsearch) (corev1.Service
 
 	return svc, nil
 }
+
+// ElasticsearchURL calculates the base url for Elasticsearch, taking into account the currently running pods.
+// If there is a HTTP scheme mismatch between spec and pods we switch to requesting individual pods directly
```
Comment (sebgl) [inline pkg/controller/elasticsearch/services/services.go]: typo: `a HTTP` -> `an HTTP`

**Comment by pebrc** [comment]
Jenkins test this please

**Comment by dzmitry-lahoda** [comment]
Any means I can get all in one yaml with this feature?

**Comment by sebgl** [comment]
@dzmitry-lahoda your best option would be to build it from the source code (`cd config/operator && make generate-all-in-one`), and also build the corresponding Docker image yourself (`make dep && docker build -t your-image-name .`). Beware the current HEAD is still considered unstable with a few bugs to fix.

Otherwise I suggest you wait a few weeks for the next release :)

**Comment by dzmitry-lahoda** [comment]
i have found CI and seems found images https://devops-ci.elastic.co/  . so seems may be ok with building yaml only. will try after installing windows subsystem for linux)

**Comment by dzmitry-lahoda** [comment]
what are guestimates for release?

**Comment by dineshabbi** [comment]
Thanks for adding this fellas ! Do I have to be on master branch to avail this ? Or does 1.0.0 have it ?

**Comment by sebgl** [comment]
@dineshabbi recently released version 1.0.0-beta1 has this feature :)

**Comment by dineshabbi** [comment]
Thanks @sebgl , it works like a charm ! Should I be aware of any settings from kibana and apm side if I want to connect to this http based elasticsearch cluster ? Can I simply ignore the settings in kibana templates like elasticsearch.ssl.certificateAuthorities: /mnt/certs/elasticsearch/ca.crt ?

**Comment by dineshabbi** [comment]
Thank you, yes that worked, appreciated the help. Is it possible to manually specify hardcoded password elastic:password in elasticsearch yaml file ? Because for whatever reason I decided to respin the cluster, I did not want to reconnect from application editing its ES_URL. Because only variable in ES_URL=http://elastic:<passwordd>@elastic_service_name:9200 is password. Perhaps I can do this via secret where I can embed a hardcoded password ? Sorry this may be a discussion topic under elastic discussion group, so will ask there if didn't get a response.

**Comment by sebgl** [comment]
If you don't care about authentication you can setup anonymous access: https://www.elastic.co/guide/en/elasticsearch/reference/7.4/anonymous-access.html

If you do care but want to specify your own password, I think you can create your own `<cluster-name>-es-elastic-user` secret, with the exact same content as the one created automatically by ECK (adapt labels accordingly).

+1 to raising a discuss thread next time :)
